src/passes/Outlining.cpp

```diff
@@ -13,23 +13,225 @@
 #define DBG(statement)
 #endif
 
+// Check a Result or MaybeResult for error and call Fatal() if the error exists.
+#define ASSERT_ERR(val)                                                        \
+  if (auto _val = (val); auto err = _val.getErr()) {                           \
+    Fatal() << err->msg;                                                       \
+  }
+
 namespace wasm {
 
+struct ReconstructStringifyWalker
+  : public StringifyWalker<ReconstructStringifyWalker> {
+
+  ReconstructStringifyWalker(Module* wasm)
+    : existingBuilder(*wasm), outlinedBuilder(*wasm) {
+    this->setModule(wasm);
+    DBG(std::cout << "\n\nexistingBuilder: " << &existingBuilder
+                  << " outlinedBuilder: " << &outlinedBuilder);
+  }
+
+  // As we reconstruct the IR during outlining, we need to know whether the
+  // expression being visited during walking is:
+  //
+  // - NotInSeq (meaning not part of a
+  // sequence that will be outlined into a new function)
+  // - InSeq (meaning part of
+  // a sequence that is being outlined into a new function)
+  // - InSkipSeq (meaning part of a sequence but has already been outlined into
+  // a new function)
+  //
+  // to know which IRBuilder to send the instruction.
+  enum ReconstructState {
+    NotInSeq = 0,
+    InSeq = 1,
+    InSkipSeq = 2,
+  };
+  ReconstructState state = ReconstructState::NotInSeq;
+  uint32_t seqCounter = 0;
+  uint32_t instrCounter = 0;
+  std::vector<OutliningSequence> sequences;
+  std::unordered_map<Name, std::vector<OutliningSequence>> seqToFunc;
+  IRBuilder existingBuilder;
+  IRBuilder outlinedBuilder;
```

Comment:
It would be good to add brief comments on each of these members variables as well.

Reply:
done